File: leekak/week19/[BOJ]1780.cpp
#include <iostream>
using namespace std;

int cnt_n=0;
int cnt_zero=0;
int cnt_p=0;
int paper[2200][2200];

// (x,y)를 좌상단으로 하고 크기 N인 정사각형이
// 한 가지 값(-1, 0, 1)으로만 이루어져 있으면 해당 값을 카운트,
// 아니면 3x3으로 나눠 9개 영역을 재귀 호출
void cut_paper(int x, int y, int N){
    bool flag=1;
    for(int i=y; i<y+N; i++) {
        for(int j=x; j<x+N; j++) {
            if(paper[y][x] != paper[i][j]) {
                flag=0; //0이면 다름
                break;
            }
        }
        if(flag==0) break;
    }
    if(flag==0) {
        cut_paper(x, y, N/3);
        cut_paper(x, y+N/3, N/3);
        cut_paper(x, y+N/3+N/3, N/3);
        cut_paper(x+N/3, y, N/3);
        cut_paper(x+N/3, y+N/3, N/3);
        cut_paper(x+N/3, y+N/3+N/3, N/3);
        cut_paper(x+N/3+N/3, y, N/3);
        cut_paper(x+N/3+N/3, y+N/3, N/3);
        cut_paper(x+N/3+N/3, y+N/3+N/3, N/3);

    } else {
        if(paper[y][x]==-1) cnt_n++;
        else if(paper[y][x]==0) cnt_zero++;
        else if(paper[y][x]==1) cnt_p++;
    }
}

int main() {
    int N;
    cin >> N;
    for(int i=0; i<N; i++) {
        for(int j=0; j<N; j++) {
            cin >> paper[i][j];
        }
    }
    cut_paper(0,0, N);
    cout << cnt_n << "\n" << cnt_zero << "\n" << cnt_p;
    return 0;
}

File: leekak/week20/[BOJ]1780.py
import sys
sys.setrecursionlimit(3000)

cnt_n = 0    # -1로 채워진 종이 개수
cnt_zero = 0 # 0으로 채워진 종이 개수
cnt_p = 0    # 1로 채워진 종이 개수
paper = []   # 종이 배열

def cut_paper(x, y, N):
    global cnt_n, cnt_zero, cnt_p

    standard_value = paper[y][x]
    is_uniform = True  # 영역이 균일한지 (하나의 값으로만 이루어졌는지) 확인

    # N*N 영역을 순회하며 확인
    for r in range(y, y + N):
        for c in range(x, x + N):
            if paper[r][c] != standard_value:
                is_uniform = False
                break
        if not is_uniform:
            break

    # 영역이 균일하지 않으면 (is_uniform == False) 9개의 작은 영역으로 분할하여 재귀 호출
    if not is_uniform:
        next_N = N // 3
        
        # 3x3 분할하여 재귀 호출
        
        cut_paper(x, y, next_N)
        cut_paper(x + next_N, y, next_N)
        cut_paper(x + 2 * next_N, y, next_N)
        
        cut_paper(x, y + next_N, next_N)
        cut_paper(x + next_N, y + next_N, next_N)
        cut_paper(x + 2 * next_N, y + next_N, next_N)
        
        cut_paper(x, y + 2 * next_N, next_N)
        cut_paper(x + next_N, y + 2 * next_N, next_N)
        cut_paper(x + 2 * next_N, y + 2 * next_N, next_N)

    # 영역이 균일하면 (is_uniform == True) 개수를 카운트
    else:
        if standard_value == -1:
            cnt_n += 1
        elif standard_value == 0:
            cnt_zero += 1
        elif standard_value == 1:
            cnt_p += 1

if __name__ == "__main__":
    try:
        N = int(sys.stdin.readline())
    except:
        N = 0
      
    for _ in range(N):
        # sys.stdin.readline()을 사용하면 한 줄 전체를 읽으므로, split()과 map(int, ...)를 사용해 리스트로 만듭니다.
        # strip()으로 개행 문자를 제거합니다.
        line = list(map(int, sys.stdin.readline().strip().split()))
        paper.append(line)

    cut_paper(0, 0, N)

    print(f"{cnt_n}\n{cnt_zero}\n{cnt_p}")
